File: sdk/python/src/basanos_sdk/consumer.py
"""Consumer-side result and event verification helpers."""
from __future__ import annotations

import struct
from dataclasses import dataclass

from solders.pubkey import Pubkey

from . import consensus as c
from .transport import RpcClient


@dataclass(frozen=True)
class ConsumerView:
    result: c.ResultV5
    events: tuple[dict, ...]


class Consumer:
    """Read DCR2 records, DLE1 events, and validate output attestations."""

    def __init__(self, rpc: RpcClient, dcg_program: Pubkey | str | bytes | bytearray) -> None:
        self.rpc = rpc
        self.dcg_program = c.key_bytes(dcg_program)

    def result_address(self, descriptor: bytes) -> bytes:
        return c.addresses(self.dcg_program, descriptor=descriptor).result[0]

    def read_result(self, descriptor: bytes) -> c.ResultV5:
        account = self.rpc.read_account(self.result_address(descriptor))
        if account is None:
            raise LookupError("result account is absent")
        if account.data[:4] == b"DCRZ":
            raise ValueError("result has been replaced by a DCRZ tombstone")
        return c.ResultV5.decode(account.data)

    def events(self, signature: str) -> list[dict]:
        return self.rpc.events_from_logs(signature, program=self.dcg_program)

    def view(self, descriptor: bytes, signature: str) -> ConsumerView:
        return ConsumerView(self.read_result(descriptor), tuple(self.events(signature)))

    def verify_output(self, data: bytes, *, position: int, segment_count: int, segment: int, entries: int,
                      local: int, region: int, offset: int, landed_root: bytes, width: int,
                      expected_table_root: bytes | None = None, expected_index: int | None = None,
                      expected_descriptor: bytes | None = None) -> dict:
        return c.verify_attestation(data, width=width, position=position, segment_count=segment_count,
                                    segment=segment, entries=entries, local=local, region=region, offset=offset,
                                    landed_root=landed_root, expected_table_root=expected_table_root,
                                    expected_index=expected_index, expected_descriptor=expected_descriptor)

    def tokens(self, result: c.ResultV5, *, vocab: int) -> list[int]:
        return usable_tokens(result, vocab=vocab)


def decode_token(value: bytes, vocab: int) -> int:
    # Bytes 0..8 hold the argmax best score (i64); only the token at 8..16 is the output.
    if len(value) != 16:
        raise ValueError(c.OUTPUT_PROOF)
    token = int.from_bytes(value[8:16], "little", signed=True)
    if not 0 <= token < vocab:
        raise ValueError(c.OUTPUT_PROOF)
    return token


def token_value(token: int) -> bytes:
    if not 0 <= token < (1 << 63):
        raise ValueError("token out of range")
    return bytes(8) + struct.pack("<q", token)


def usable_tokens(result: c.ResultV5, *, vocab: int) -> list[int]:
    if result.status not in c.USABLE_STATUSES or result.document_root == bytes(32):
        raise ValueError("result is not usable")
    if result.outputs is None or result.outputs_attested != result.output_count:
        raise ValueError(c.OUTPUT_PROOF)
    return [decode_token(value, vocab) for value in result.outputs if value is not None]

File: sdk/typescript/src/consumer.ts
import { PublicKey } from "@solana/web3.js";
import * as c from "./consensus.js";
import { RpcClient } from "./transport.js";

export interface ConsumerView {
  result: c.ResultV5;
  events: ReadonlyArray<Record<string, c.Integer | Buffer | string>>;
}

export class Consumer {
  readonly rpc: RpcClient;
  readonly dcgProgram: PublicKey;

  constructor(rpc: RpcClient, dcgProgram: PublicKey | string | Uint8Array) {
    this.rpc = rpc;
    this.dcgProgram = c.key(dcgProgram);
  }

  resultAddress(descriptor: Uint8Array): PublicKey {
    const value = c.addresses(this.dcgProgram, { descriptor }).result;
    if (value === undefined) throw new Error("missing result address");
    return value[0];
  }

  result_address(descriptor: Uint8Array): PublicKey { return this.resultAddress(descriptor); }

  async readResult(descriptor: Uint8Array): Promise<c.ResultV5> {
    const account = await this.rpc.readAccount(this.resultAddress(descriptor));
    if (account === null) throw new Error("result account is absent");
    if (account.data.subarray(0, 4).equals(Buffer.from("DCRZ"))) throw new Error("result has been replaced by a DCRZ tombstone");
    return c.ResultV5.decode(account.data);
  }

  async read_result(descriptor: Uint8Array): Promise<c.ResultV5> { return this.readResult(descriptor); }

  async events(signature: string): Promise<Array<Record<string, c.Integer | Buffer | string>>> {
    return this.rpc.eventsFromLogs(signature, this.dcgProgram);
  }

  async view(descriptor: Uint8Array, signature: string): Promise<ConsumerView> {
    return { result: await this.readResult(descriptor), events: await this.events(signature) };
  }

  verifyOutput(data: Uint8Array, options: { position: number; segmentCount: number; segment: number; entries: number; local: number; region: number; offset: number; landedRoot: Uint8Array; width: number; expectedTableRoot?: Uint8Array; expectedIndex?: number; expectedDescriptor?: Uint8Array }): c.AttestationVerification {
    return c.verifyAttestation(data, options);
  }

  verify_output(data: Uint8Array, options: { position: number; segmentCount: number; segment: number; entries: number; local: number; region: number; offset: number; landedRoot: Uint8Array; width: number; expectedTableRoot?: Uint8Array; expectedIndex?: number; expectedDescriptor?: Uint8Array }): c.AttestationVerification {
    return this.verifyOutput(data, options);
  }

  tokens(result: c.ResultV5, options: { vocab: number }): number[] { return usableTokens(result, options); }
}

export function decodeToken(value: Uint8Array, vocab: number): number {
  // Bytes 0..8 hold the argmax best score (i64); only the token at 8..16 is the output.
  if (value.length !== 16) throw new Error(String(c.OUTPUT_PROOF));
  const raw = Buffer.from(value.subarray(8, 16)).readBigUInt64LE(0);
  const token = BigInt.asIntN(64, raw);
  if (token < 0n || token >= BigInt(vocab)) throw new Error(String(c.OUTPUT_PROOF));
  return Number(token);
}

export const decode_token = decodeToken;

export function tokenValue(token: number | bigint): Buffer {
  const value = typeof token === "bigint" ? token : BigInt(token);
  if (value < 0n || value >= 1n << 63n) throw new Error("token out of range");
  const out = Buffer.alloc(16);
  out.writeBigUInt64LE(value, 8);
  return out;
}

export const token_value = tokenValue;

export function usableTokens(result: c.ResultV5, options: { vocab: number }): number[] {
  if (!c.USABLE_STATUSES.has(result.status) || result.documentRoot.length !== 32 || result.documentRoot.every((item) => item === 0)) throw new Error("result is not usable");
  if (result.outputs.length !== result.outputCount || result.outputsAttested !== result.outputCount) throw new Error(String(c.OUTPUT_PROOF));
  return result.outputs.map((value) => {
    if (value === null) throw new Error(String(c.OUTPUT_PROOF));
    return decodeToken(value, options.vocab);
  });
}

export const usable_tokens = usableTokens;
